src/audio/tone.c

```diff
@@ -612,17 +618,22 @@ static int tone_copy(struct comp_dev *dev)
 {
 	struct comp_buffer *sink;
 	struct comp_data *cd = comp_get_drvdata(dev);
+	uint32_t free;
 
 	comp_dbg(dev, "tone_copy()");
 
 	/* tone component sink buffer */
 	sink = list_first_item(&dev->bsink_list, struct comp_buffer,
 			       source_list);
 
+	buffer_lock(sink);
+	free = sink->stream.free;
+	buffer_unlock(sink);
```

Comment:
Another buffer_ API call ?
```
buffer_get_stream_free()
```
Probably a few other buffer calls could benefit here with locking ?

Reply:
Locking it on such granular level isn't efficient and will lead to deadlocks. As I stated in the previous comment, we need to think in the near future about some kind of additional wrappers for getting/releasing buffers to avoid explicit lock/unlock calls. Then the code will look much more cleaner and it will be easier for the integrators to implement their own processing components.

Reply:
Maybe the wrapper copy() code could take the lock for both buffers in some atomic fashion (ordering may be sufficient, first source then sink) and then call the specific copy callback, which doesn't need to worry about this at all?

Reply:
@paulstelian97 That was my initial idea, but we can't block the whole copy. It's making the idea of multicore processing useless.

Reply:
Alright, but I don't think there is a single recipe which will work for all components either.

For my future use case, where I have processing functions that expect linear buffers (don't know how to interact with the SOF buffers and are closed source so I cannot update them), what would be the best recipe for locking and unlocking buffers? Would there be a need to allocate an additional buffer for that situation?

Reply:
I don't think we have to block the whole copy in this use case. The only locking and atomic code should be around R/W pointer updates and free/avail updates. Nothing else.

Reply:
So the buffer_update_*, buffer_consume_* and buffer_produce_* could have an internal lock? We have to assume there is only one producer and one consumer per buffer (it can be established and isn't a real limitation I guess).

Yup, this can work.

Reply:
@lgirdwood It's done exactly this way now. Would like to also explicitly block irq?

Reply:
@tlauda ok great, I wasn't sure from the PR diffs, yes I think the free/avail calc should be atomic (and block IRQs).

Reply:
@lgirdwood Done.